extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go

```diff
@@ -213,70 +214,49 @@ func (v *vpcFlowLogUnmarshaler) addToLogs(
 		}
 	}
 
-	// get the address fields
-	addresses := v.handleAddresses(ips)
-	for field, value := range addresses {
-		record.Attributes().PutStr(field, value)
+	if logLine != "" {
+		return errors.New("log line has more fields than the ones expected")
 	}
 
-	scopeLogs := v.getScopeLogs(*key, scopeLogsByResource)
+	// add the address fields with the correct conventions
+	// to the log record
+	v.handleAddresses(addr, record)
 	rScope := scopeLogs.LogRecords().AppendEmpty()
 	record.MoveTo(rScope)
 
 	return nil
 }
 
-// handleAddresses creates a new map where the original field
-// names will be the known conventions for the fields
-func (v *vpcFlowLogUnmarshaler) handleAddresses(addresses map[string]string) map[string]string {
-	// max is 3 fields, see example in
+// handleAddresses creates adds the addresses to the log record
+func (v *vpcFlowLogUnmarshaler) handleAddresses(addr *address, record plog.LogRecord) {
+	// see example in
 	// https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs-records-examples.html#flow-log-example-nat
-	recordAttr := make(map[string]string, 3)
-	srcaddr, foundSrc := addresses["srcaddr"]
-	pktSrcaddr, foundSrcPkt := addresses["pkt-srcaddr"]
-	if !foundSrcPkt && foundSrc {
+	if addr.pktSource == "" && addr.source != "" {
 		// there is no middle layer, assume "srcaddr" field
 		// corresponds to the original source address.
-		recordAttr[conventions.AttributeSourceAddress] = srcaddr
-	} else if foundSrcPkt && foundSrc {
-		recordAttr[conventions.AttributeSourceAddress] = pktSrcaddr
-		if srcaddr != pktSrcaddr {
+		record.Attributes().PutStr(conventions.AttributeSourceAddress, addr.source)
+	} else if addr.pktSource != "" && addr.source != "" {
+		record.Attributes().PutStr(conventions.AttributeSourceAddress, addr.pktSource)
+		if addr.pktSource != addr.source {
 			// srcaddr is the middle layer
-			recordAttr[conventions.AttributeNetworkPeerAddress] = srcaddr
+			record.Attributes().PutStr(conventions.AttributeNetworkLocalAddress, addr.source)
 		}
 	}
 
-	dstaddr, foundDst := addresses["dstaddr"]
-	pktDstaddr, foundDstPkt := addresses["pkt-dstaddr"]
-	if !foundDstPkt && foundDst {
+	if addr.pktDestination == "" && addr.destination != "" {
 		// there is no middle layer, assume "dstaddr" field
 		// corresponds to the original destination address.
-		recordAttr[conventions.AttributeDestinationAddress] = dstaddr
-	} else if foundDstPkt && foundDst {
-		recordAttr[conventions.AttributeDestinationAddress] = pktDstaddr
-		if pktDstaddr != dstaddr {
-			if _, found := recordAttr[conventions.AttributeNetworkPeerAddress]; found {
+		record.Attributes().PutStr(conventions.AttributeDestinationAddress, addr.destination)
+	} else if addr.pktDestination != "" && addr.destination != "" {
+		record.Attributes().PutStr(conventions.AttributeDestinationAddress, addr.pktDestination)
+		if addr.pktDestination != addr.destination {
+			if _, found := record.Attributes().Get(conventions.AttributeNetworkPeerAddress); found {
```

Comment:
Instead of checking for record attributes (this does a linear scan), can you add a variable like `var setNetworkLocalAddress bool` which we set to true in the source block above? Then check that here.

Reply:
Yes. I have added a bool variable there now, instead of the `getStr`.